### PR TITLE
Optimize ContainerTop() aka docker top

### PR DESCRIPTION
**Pull request by kolyshkin**
**- What I did**

Optimized the implementation of `ContainerTop()` (aka `docker top`) resulting in 5x to 100x speed improvement (and some savings in CPU and memory usage that were not measured).

**- How I did it**

Current ContainerTop (a.k.a. docker top) implementation uses "ps" to get the info about *all* running processes, then parses it, then    filters the results to only contain PIDs used by the container.    Collecting data only to throw most of it away is inefficient,    especially on a system running many containers (or processes).    For example, "docker top" on a container with a single process    can take up to 0.5 seconds to execute (on a mostly idle system)    which is noticeably slow.
    
Since the containers PIDs are known beforehand, let's use ps's    "q" option to provide it with a list of PIDs we want info about.

**- How to verify it**

1. Existing integration test case `TestContainerAPITop`.
2. Manual run of `docker top` with various options.

**- Description for the changelog**
* ContainerTop speed up

**- A picture of a cute animal (not mandatory but encouraged)**

Moscow zoo racoon
Source: https://commons.wikimedia.org/wiki/File:Raccoon_Moscow_zoo.JPG

### CODE REVIEW

File: daemon/top_unix.go [hunk 1]
```diff
@@ -144,9 +161,23 @@ func (daemon *Daemon) ContainerTop(name string, psArgs string) (*container.Conta
...
+		output, err = exec.Command("ps", args...).Output()
+		if err != nil {
+			if ee, ok := err.(*exec.ExitError); ok {
+				// first line of stderr shows why ps failed
+				line := bytes.SplitN(ee.Stderr, []byte{'\n'}, 2)
+				if len(line) > 0 && len(line[0]) > 0 {
+					err = errors.New(string(line[0]))
+				}
+			}
+			return nil, fmt.Errorf("Error running ps: %v", err)
```
Comment (cpuguy83) [inline daemon/top_unix.go]: `errdefs.System(errors.Wrap(err, "error running ps"))`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
```
01:26:00 /docker-py/tests/integration/api_container_test.py:1102: in test_top
01:26:00     assert res['Titles'] == [
01:26:00 E   AssertionError: assert ['UID', 'PID'...', 'TTY', ...] == ['UID', 'PID',...', 'TTY', ...]
01:26:00 E     At index 6 diff: u'STAT' != 'TIME'
01:26:00 E     Left contains more items, first extra item: 'CMD'
01:26:00 E     Use -v to get the full diff
01:26:00 === 1 failed, 327 passed, 5 skipped, 4 xfailed, 1 xpassed in 388.61 seconds ====
```

File: daemon/top_unix.go (hunk 1 above)
Comment (kolyshkin) [reply daemon/top_unix.go]: done

**Comment by kolyshkin** [comment]
As for the docker-py test failure, indeed adding `q` to ps for some reason makes an extra column, `STAT`, appear in the output:
```
$ ps -ef q 1 | head -2
UID        PID  PPID  C STIME TTY      STAT   TIME CMD
root         1     0  0 May12 ?        Ss     0:43 /sbin/init splash
$ ps -ef | head -2
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 May12 ?        00:00:43 /sbin/init splash
```

I'll take a look as to why...

**Comment by kolyshkin** [comment]
OK, the docker-py CI test failure should now be fixed by replacing `q` with `-q`. This breaks using `busybox ps` but AFAICS it can not be used anyway.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
> This breaks using busybox ps but AFAICS it can not be used anyway.

May actually be an issue on Docker for Mac?

**Review by boaz0 — APPROVED** [review]

**Comment by kolyshkin** [comment]
> May actually be an issue on Docker for Mac?

Frankly I have no idea, but in general it looks like we rely on having "proper" ps in a few places.

Is there a way to test it (without having a Mac)?

**Comment by thaJeztah** [comment]
> Is there a way to test it (without having a Mac)?

Hm, good one. Not sure.

Did realise that the official Docker images on Hub could have that issue as well (as they're Alpine based, using busybox)

**Comment by thaJeztah** [comment]
Seems to be working; could it be dependent on which version of Busybox?

```bash
make BIND_DIR=. binary
cd bundles/binary-daemon

docker build -t testing . -f-<<EOF
FROM docker:18.05-dind
COPY /dockerd-dev /usr/local/bin/dockerd
EOF

docker run --name dockie -d --rm --privileged -v /var/lib/docker testing

docker exec -it dockie sh
/ # docker run -d --name webbie nginx:alpine
/ # docker top webbie
PID                 USER                TIME                COMMAND
241                 root                0:00                nginx: master process nginx -g daemon off;
285                 dockrema            0:00                nginx: worker process
```

```
Client:
 Version:      18.05.0-ce
 API version:  1.37
 Go version:   go1.9.2
 Git commit:   f150324
 Built:        Wed May  9 22:11:29 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      dev
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   a41328d570
  Built:        Tue May 29 10:26:34 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by thaJeztah** [comment]
Odd, because running it directly inside the container fails;

```
 ps -ef -q 1 | head -2
ps: unrecognized option: q
BusyBox v1.27.2 (2017-12-12 10:41:50 GMT) multi-call binary.

Usage: ps [-o COL1,COL2=HEADER]

Show list of processes

	-o COL1,COL2=HEADER	Select columns for display
```

**Comment by kolyshkin** [comment]
This commit retries without `-q` if it fails. I wanted to avoid this retry for busybox by using `q` (instead of `-q`) but it broke the docker-py test because non-busybox `ps -ef q<PID>,..` adds one extra STAT column (which docker-py does not expect). There are two ways to solve it:
1. Use `-q` (means always retry for busybox).
2. Use `q` (means failing the docker-py test case -- and quite possibly some other users of `docker top`).

The original version of this patch set used approach No. 2, when I switched to No. 1.

I think I can amend this by remembering we're on a platform that fails `ps -ef -q` and so the retry won't happen (for the rest of dockerd process runtime, that is). Let me see if I can do it in a non-ugly way.

**Review by thaJeztah — APPROVED** [review]
works for me; let's do this! LGTM

**Comment by kolyshkin** [comment]
> Let me see if I can do it in a non-ugly way.

Whatever I do it looks kinda ugly. I guess we may leave it as a followup.

**Pull request merged**